### PR TITLE
Integrate TypeScript

### PR DESCRIPTION
**Pull request by pbelbin**
Adds dist folder into which the TypeScript compiler will put converted src files.

Note: test content are not to be written in TypeScript currently, so they are not converted, but, the tests themselves pull in the converted javascript (from dist) and perform the tests.

ESLint config has also been modified to support typescript source and rules.

Because generated files are put into the dist folder, I added a clean script to wipe out any such generated files (including the coverage folder contents. Note: this assumes you're using a unix flavored environment, as it won't work in Windows environments.

I also added a build-ts script, but, imagine it won't get much use in this instance. Might want to remove that.

Yes, there are a number of dependencies that are not entirely necessary right now, since there's no actual typescript source files. But, it should be ready now to add them.

### CODE REVIEW

**Comment by pbelbin** [comment]
Adds `dist` folder into which the TypeScript compiler will put converted `src` files.

Note: `test` content are *not* to be written in TypeScript currently, so they are not converted, but, the tests themselves pull in the converted javascript (from `dist`) and perform the tests.

ESLint config has also been modified to support typescript source and rules.

Because generated files are put into the `dist` folder, I added a `clean` script to wipe out any such generated files (including the `coverage` folder contents.  Note: this assumes you're using a unix flavored environment, as it won't work in Windows environments.

I also added a `build-ts` script, but, imagine it won't get much use in this instance.  Might want to remove that.

File: .eslintrc.json [hunk 1]
```diff
@@ -1,19 +1,34 @@
 {
-	"extends": ["eslint:recommended", "prettier"],
 	"env": {
 		"es6": true,
 		"node": true
 	},
+	"extends": [
```
Comment (WesCossick) [inline .eslintrc.json]: Let's keep the `extends` key the first one listed in the `.eslintrc.json` file.

File: .eslintrc.json [hunk 2]
```diff
@@ -1,19 +1,34 @@
...
+		"eslint:recommended",
+		"prettier",
```
Comment (WesCossick) [inline .eslintrc.json]: The two `prettier` extends need to be the final ones. That way, they can overwrite anything that comes before them.

File: .eslintrc.json [hunk 3]
```diff
@@ -1,19 +1,34 @@
...
 	"parserOptions": {
 		"ecmaVersion": 2018,
-		"sourceType": "module"
+		"sourceType": "module",
+		"project": ["./tsconfig.json", "./tsconfig.eslint.json"]
 	},
+	"plugins": ["@typescript-eslint"],
 	"rules": {
-		"no-unused-vars": [
+		"no-unused-vars": "off",
```
Comment (WesCossick) [inline .eslintrc.json]: `no-unused-vars` still needs to be an error.

File: package.json [hunk 4]
```diff
@@ -2,14 +2,16 @@
...
 	"files": [
-		"/src"
+		"/dist"
 	],
 	"scripts": {
-		"test": "jest",
-		"lint": "eslint . --ext .js && prettier --check '**/*.{js,json}'",
-		"format": "eslint --fix . --ext .js && prettier --write '**/*.{js,json}'"
+		"test": "tsc && jest",
+		"lint": "eslint --ext .js,.ts,.tsx src test && prettier --check '**/*.{ts,tsx,js,json}'",
```
Comment (WesCossick) [inline package.json]: `.tsx` is only for React projects, which this is not.

File: package.json [hunk 5]
```diff
@@ -2,14 +2,16 @@
...
+		"format": "eslint --fix --ext .js,.ts,.tsx src test && prettier --write '**/*.{ts,tsx,js,json}'",
+		"build-ts": "tsc",
```
Comment (WesCossick) [inline package.json]:
```suggestion
		"compile": "tsc",
```

File: package.json [hunk 6]
```diff
@@ -2,14 +2,16 @@
...
+		"clean": "rm -rf dist coverage"
```
Comment (WesCossick) [inline package.json]: This probably shouldn't be a command in the package file.

File: package.json [hunk 7]
```diff
@@ -26,22 +28,38 @@
 	},
 	"homepage": "https://github.com/sparksuite/waterfall-cli",
 	"dependencies": {
+		"@types/node": "^12.7.7",
```
Comment (WesCossick) [inline package.json]: Shouldn't this be a `devDependencies`?

File: package.json [hunk 8]
```diff
@@ -26,22 +28,38 @@
...
+		"prettier": "1.18.2",
+		"ts-jest": "^24.1.0",
+		"ts-node": "^8.4.1",
+		"typescript": "^3.6.3"
 	},
 	"jest": {
 		"verbose": true,
 		"coverageDirectory": "./coverage/",
-		"collectCoverage": true
+		"collectCoverage": true,
```
Comment (WesCossick) [inline package.json]: I don't think any configuration changes are necessary to Jest.

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
+	"compilerOptions": {
+		"target": "es6",
```
Comment (WesCossick) [inline tsconfig.json]: Can you sort these in alphabetical order?

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
+{
+	"compilerOptions": {
```
Comment (WesCossick) [inline tsconfig.json]: We _may_ need `"resolveJsonModule": true,` in there too, based on the TypeScript I've been writing in another project.

File: .eslintrc.json [hunk 11]
```diff
@@ -1,19 +1,34 @@
...
+		"@typescript-eslint/no-var-requires": "off",
+		"@typescript-eslint/unbound-method": "off",
+		"@typescript-eslint/prefer-regexp-exec": "off",
+		"@typescript-eslint/no-unused-vars": [
 			"error",
 			{
 				"vars": "all",
-				"args": "after-used"
+				"args": "after-used",
+				"ignoreRestSiblings": false
```
Comment (WesCossick) [inline .eslintrc.json]: This doesn't need to be included, it appears `false` is the default value already.

**Comment by pbelbin** [comment]
Problem: If, after a build is done by TypeScript (even with the --watch option), the source for that output file is removed, the compiled file remains.  With the --watch option, TypeScript will notice and perform the compile of a new or changed file, but, not the removal.

I suppose this is analogous to many other compilers.

There is a command sequence which will completely empty the output folder:

```
tsc --build --clean
```

(it's a sub-command of the 'build', it seems)

And then, the build is needed again, to populate the output folder.

**Pull request merged**